File: app/test/question_status_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:auto_scoring_app/api/sidecar_api_client.dart';
import 'package:auto_scoring_app/core/design/app_status_tone.dart';
import 'package:auto_scoring_app/core/question_status.dart';

JobResponse _job({
  String state = 'running',
  bool? usable,
  String? blockedOnQuestionId,
  String questionId = 'q1',
  String? id,
  int createdAtSeconds = 0,
}) => JobResponse(
  (b) => b
    ..id = id ?? 'job-$questionId'
    ..kind = 'grading'
    ..submissionId = 'sub-1'
    ..questionId = questionId
    ..state = state
    ..usable = usable
    ..blockedOnQuestionId = blockedOnQuestionId
    ..attempts = 1
    ..maxAttempts = 3
    ..createdAt = DateTime.utc(
      2026,
      1,
      1,
    ).add(Duration(seconds: createdAtSeconds))
    ..updatedAt = DateTime.utc(
      2026,
      1,
      1,
    ).add(Duration(seconds: createdAtSeconds)),
);

ReviewResponse _review({
  String action = 'approved',
  String? regradeJobId,
  int createdAtSeconds = 0,
}) => ReviewResponse(
  (b) => b
    ..id = 'review-1'
    ..submissionId = 'sub-1'
    ..questionId = 'q1'
    ..action = action
    ..version = 1
    ..aiGradeResultId = 'grade-1'
    ..regradeJobId = regradeJobId
    ..createdAt = DateTime.utc(
      2026,
      1,
      1,
    ).add(Duration(seconds: createdAtSeconds)),
);

void main() {
  group('deriveQuestionStatus', () {
    test('has no job and no review: 未処理', () {
      expect(
        deriveQuestionStatus(job: null, review: null),
        QuestionStatus.pending,
      );
    });

    test('maps each queue state to its own node state', () {
      for (final (state, expected) in const [
        ('blocked', QuestionStatus.blocked),
        ('queued', QuestionStatus.queued),
        ('running', QuestionStatus.running),
        ('failed', QuestionStatus.failed),
        ('cancelled', QuestionStatus.cancelled),
      ]) {
        expect(
          deriveQuestionStatus(job: _job(state: state), review: null),
          expected,
          reason: state,
        );
      }
    });

    test('a live job outranks an older review decision', () {
      // Re-submission under a newer graph version, or a 再判定 request,
      // creates a fresh job while the append-only review history still holds
      // the previous attempt's 承認 -- showing 承認済み over it would say the
      // opposite of what is happening.
      expect(
        deriveQuestionStatus(
          job: _job(state: 'running'),
          review: _review(),
        ),
        QuestionStatus.running,
      );
    });

    test('succeeded but not usable is 要確認, not レビュー待ち', () {
      // The queue itself judged the result untrustworthy and is holding
      // everything downstream (docs/job-queue.md) -- a stronger statement
      // than "nobody has reviewed it yet".
      expect(
        deriveQuestionStatus(
          job: _job(state: 'succeeded', usable: false),
          review: null,
        ),
        QuestionStatus.needsCheck,
      );
    });

    test('a succeeded job carries the human decision when there is one', () {
      for (final (action, expected) in const [
        (null, QuestionStatus.graded),
        ('approved', QuestionStatus.approved),
        ('modified', QuestionStatus.approved),
        ('rejected', QuestionStatus.rejected),
        ('regrade_requested', QuestionStatus.regradeRequested),
        ('undone', QuestionStatus.graded),
      ]) {
        expect(
          deriveQuestionStatus(
            job: _job(state: 'succeeded', usable: true),
            review: action == null ? null : _review(action: action),
          ),
          expected,
          reason: '$action',
        );
      }
    });

    test(
      'a stopped job does not outrank a decision made since (Issue #118)',
      () {
        // A job that failed, was cancelled, or finished with an untrustworthy
        // result is not going to describe anything newer. A person who has
        // graded the question by hand since then is the newer fact, and the
        // answer sheet already counts that question as 確認済み server-side --
        // leaving 失敗 here would have the rail, the DAG and the Inspector all
        // contradicting the submission's own state.
        for (final (state, usable) in const [
          ('failed', false),
          ('cancelled', null),
          ('succeeded', false),
        ]) {
          expect(
            deriveQuestionStatus(
              job: _job(state: state, usable: usable, createdAtSeconds: 1),
              review: _review(action: 'modified', createdAtSeconds: 2),
            ),
            QuestionStatus.approved,
            reason: state,
          );
        }
      },
    );

    test('a stopped job still outranks a decision that predates it', () {
      // The other half of the same rule: a re-submission creates a fresh job
      // after an old 承認, and that approval is about the previous attempt.
      for (final (state, usable, expected) in const [
        ('failed', false, QuestionStatus.failed),
        ('cancelled', null, QuestionStatus.cancelled),
        ('succeeded', false, QuestionStatus.needsCheck),
      ]) {
        expect(
          deriveQuestionStatus(
            job: _job(state: state, usable: usable, createdAtSeconds: 2),
            review: _review(action: 'modified', createdAtSeconds: 1),
          ),
          expected,
          reason: state,
        );
      }
    });

    test('再判定待ち ends when the job the request created succeeds', () {
      // Nothing ever closes a `regrade_requested` row -- the replacement
      // grade lands as a new GradeResult, not a new Review -- so treating
      // the action as the state left the node stuck on 再判定待ち with a
      // fresh grade on screen unmentioned (review round 1, P2).
      final review = _review(
        action: 'regrade_requested',
        regradeJobId: 'job-regrade',
      );
      expect(
        deriveQuestionStatus(
          job: _job(state: 'succeeded', usable: true, id: 'job-regrade'),
          review: review,
        ),
        QuestionStatus.graded,
      );
      // Still outstanding while only the *superseded* attempt is known.
      expect(
        deriveQuestionStatus(
          job: _job(state: 'succeeded', usable: true, id: 'job-original'),
          review: review,
        ),
        QuestionStatus.regradeRequested,
      );
    });

    test('再判定待ち also ends for any job that ran after the request', () {
      // The replacement job is normally named by the review, but a job
      // created afterwards for another reason (a re-submission under a newer
      // graph version) did not produce the result the reviewer rejected
      // either.
      final review = _review(action: 'regrade_requested', createdAtSeconds: 10);
      expect(
        deriveQuestionStatus(
          job: _job(state: 'succeeded', usable: true, createdAtSeconds: 20),
          review: review,
        ),
        QuestionStatus.graded,
      );
      expect(
        deriveQuestionStatus(
          job: _job(state: 'succeeded', usable: true, createdAtSeconds: 5),
          review: review,
        ),
        QuestionStatus.regradeRequested,
      );
    });
  });

  group('QuestionStatus presentation', () {
    test('every state has its own icon and label, not just a colour', () {
      // Issue #25's rule: colour only sharpens a distinction that already
      // survives without it.
      final labels = QuestionStatus.values.map((s) => s.label).toSet();
      final icons = QuestionStatus.values.map((s) => s.icon).toSet();
      expect(labels, hasLength(QuestionStatus.values.length));
      expect(icons, hasLength(QuestionStatus.values.length));
    });

    test('only 要確認 pulls the eye, and only a failure is danger', () {
      final attention = QuestionStatus.values
          .where((s) => s.tone == AppStatusTone.attention)
          .toSet();
      final danger = QuestionStatus.values
          .where((s) => s.tone == AppStatusTone.danger)
          .toSet();
      expect(attention, {QuestionStatus.needsCheck});
      expect(danger, {QuestionStatus.failed});
    });
  });
}

File: desktop/test/question-status.test.ts
import { describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import path from "node:path";

import type { components } from "../src/renderer/api/generated/schema.js";
import {
  deriveQuestionStatus,
  jobIsInProgress,
  latestJobFor,
} from "../src/renderer/core/question-status.js";

type JobResponse = components["schemas"]["JobResponse"];
type ReviewResponse = components["schemas"]["ReviewResponse"];

const PACKAGE_ROOT = path.resolve(import.meta.dirname, "..");

function atUtc(seconds: number): string {
  return new Date(Date.UTC(2026, 1, 1, 0, 0, seconds)).toISOString();
}

function job(
  overrides: Partial<JobResponse> & {
    state?: string;
    usable?: boolean | null;
    createdAtSeconds?: number;
  } = {},
): JobResponse {
  const { createdAtSeconds = 0, usable, ...rest } = overrides;
  return {
    id: "job-q1",
    kind: "grading",
    submission_id: "sub-1",
    question_id: "q1",
    state: "running",
    blocked_on_question_id: null,
    attempts: 1,
    max_attempts: 3,
    created_at: atUtc(createdAtSeconds),
    updated_at: atUtc(createdAtSeconds),
    ...rest,
    ...(usable !== undefined ? { usable } : {}),
  };
}

function review(
  overrides: Partial<ReviewResponse> & {
    action?: string;
    createdAtSeconds?: number;
  } = {},
): ReviewResponse {
  const createdAtSeconds = overrides.createdAtSeconds ?? 0;
  return {
    id: overrides.id ?? "review-1",
    submission_id: "sub-1",
    question_id: overrides.question_id ?? "q1",
    action: overrides.action ?? "approved",
    version: 1,
    ai_grade_result_id: "grade-1",
    regrade_job_id: overrides.regrade_job_id ?? null,
    created_at: atUtc(createdAtSeconds),
    ...overrides,
  };
}

function derive(input: {
  job: JobResponse | null | undefined;
  review: ReviewResponse | null | undefined;
  hasWaitingDependents?: boolean;
}) {
  return deriveQuestionStatus({
    job: input.job,
    review: input.review,
    hasWaitingDependents: input.hasWaitingDependents ?? false,
  });
}

describe("deriveQuestionStatus", () => {
  describe("INV-152: usable=false branches on whether dependents are waiting", () => {
    it("returns needsCheck only when unusable and a dependent is blocked behind it", () => {
      const unusable = job({ state: "succeeded", usable: false });
      expect(
        derive({ job: unusable, review: null, hasWaitingDependents: true }),
      ).toBe("needsCheck");
      expect(
        derive({ job: unusable, review: null, hasWaitingDependents: false }),
      ).toBe("graded");
    });

    it("does not treat usable alone as needsCheck when nothing is waiting", () => {
      for (const usable of [true, false] as const) {
        expect(
          derive({
            job: job({ state: "succeeded", usable }),
            review: null,
            hasWaitingDependents: false,
          }),
        ).toBe("graded");
      }
    });

    it("keeps an approval recorded before the attempt when nothing is waiting", () => {
      for (const usable of [true, false] as const) {
        expect(
          derive({
            job: job({
              state: "succeeded",
              usable,
              createdAtSeconds: 2,
            }),
            review: review({ action: "approved", createdAtSeconds: 1 }),
            hasWaitingDependents: false,
          }),
        ).toBe("approved");
      }
    });
  });

  describe("INV-153: a human decision after a stopped job outranks the job state", () => {
    it("shows the review when it was recorded after the terminal job", () => {
      for (const state of ["failed", "cancelled", "succeeded"] as const) {
        expect(
          derive({
            job: job({
              state,
              usable: state === "succeeded" ? false : null,
              createdAtSeconds: 1,
            }),
            review: review({ action: "modified", createdAtSeconds: 2 }),
            hasWaitingDependents: true,
          }),
        ).toBe("approved");
      }
    });

    it("keeps the stopped job when the review predates it", () => {
      expect(
        derive({
          job: job({ state: "failed", usable: false, createdAtSeconds: 2 }),
          review: review({ action: "modified", createdAtSeconds: 1 }),
          hasWaitingDependents: true,
        }),
      ).toBe("failed");
      expect(
        derive({
          job: job({ state: "cancelled", createdAtSeconds: 2 }),
          review: review({ action: "modified", createdAtSeconds: 1 }),
          hasWaitingDependents: true,
        }),
      ).toBe("cancelled");
      expect(
        derive({
          job: job({
            state: "succeeded",
            usable: false,
            createdAtSeconds: 2,
          }),
          review: review({ action: "modified", createdAtSeconds: 1 }),
          hasWaitingDependents: true,
        }),
      ).toBe("needsCheck");
    });

    it("still lets a live job outrank an older review", () => {
      expect(
        derive({
          job: job({ state: "running", createdAtSeconds: 2 }),
          review: review({ action: "approved", createdAtSeconds: 1 }),
        }),
      ).toBe("running");
    });
  });
});

describe("latestJobFor (Issue #402)", () => {
  it("returns null when the question has no job at all", () => {
    expect(latestJobFor([], "q1")).toBeNull();
    expect(latestJobFor([job({ question_id: "q2" })], "q1")).toBeNull();
  });

  it("returns the most recently created job, not the first in the list", () => {
    const oldJob = job({ id: "job-old", createdAtSeconds: 1 });
    const regradeJob = job({ id: "job-regrade", createdAtSeconds: 2 });
    // The API returns created_at ascending, so the first row is the oldest.
    expect(latestJobFor([oldJob, regradeJob], "q1")?.id).toBe("job-regrade");
    // Order in the list must not change the answer.
    expect(latestJobFor([regradeJob, oldJob], "q1")?.id).toBe("job-regrade");
  });

  it("ignores other questions' jobs and keeps the first row on a tie", () => {
    const first = job({ id: "job-first", question_id: "q1" });
    const tied = job({ id: "job-tied", question_id: "q1" });
    // Newer, but about another question: it must not win.
    const other = job({
      id: "job-other",
      question_id: "q2",
      createdAtSeconds: 5,
    });
    expect(latestJobFor([first, other, tied], "q1")?.id).toBe("job-first");
  });
});

describe("Issue #402: a regrade is answered by its own job", () => {
  it("reads 再判定待ち only until the newest job answers the request", () => {
    const oldJob = job({
      id: "job-old",
      state: "succeeded",
      usable: true,
      createdAtSeconds: 1,
    });
    const regradeJob = job({
      id: "job-regrade",
      state: "succeeded",
      usable: true,
      createdAtSeconds: 2,
    });
    const request = review({
      action: "regrade_requested",
      regrade_job_id: "job-regrade",
      createdAtSeconds: 1,
    });

    // Passing the superseded attempt is what left the screen stuck.
    expect(derive({ job: oldJob, review: request })).toBe("regradeRequested");
    // The newest job answers the request, so the question is reviewable again.
    expect(derive({ job: regradeJob, review: request })).toBe("graded");
    expect(
      derive({
        job: latestJobFor([oldJob, regradeJob], "q1"),
        review: request,
      }),
    ).toBe("graded");
  });

  it("shows the replacement attempt while it is still running", () => {
    const oldJob = job({
      id: "job-old",
      state: "succeeded",
      usable: true,
      createdAtSeconds: 1,
    });
    const running = job({
      id: "job-regrade",
      state: "running",
      usable: null,
      createdAtSeconds: 2,
    });
    const request = review({
      action: "regrade_requested",
      regrade_job_id: "job-regrade",
      createdAtSeconds: 1,
    });
    expect(
      derive({ job: latestJobFor([oldJob, running], "q1"), review: request }),
    ).toBe("running");
  });
});

describe("INV-006: question status is derived centrally", () => {
  it("only tells a caller a job is in flight for non-terminal states (Issue #319)", () => {
    expect(jobIsInProgress(null)).toBe(false);
    expect(jobIsInProgress(undefined)).toBe(false);
    for (const state of ["queued", "running", "blocked"]) {
      expect(jobIsInProgress(job({ state }))).toBe(true);
    }
    for (const state of ["succeeded", "failed", "cancelled"]) {
      expect(jobIsInProgress(job({ state }))).toBe(false);
    }
  });

  const STATUS_SURFACES = [
    "src/renderer/features/pdf-review/PdfReviewPage.tsx",
    "src/renderer/features/submission-confirm/SubmissionConfirmPage.tsx",
    "src/renderer/core/dependency-dag.ts",
  ] as const;

  it("lists every surface that shows a per-question status", () => {
    expect(STATUS_SURFACES).toEqual([
      "src/renderer/features/pdf-review/PdfReviewPage.tsx",
      "src/renderer/features/submission-confirm/SubmissionConfirmPage.tsx",
      "src/renderer/core/dependency-dag.ts",
    ]);
  });

  it("routes each listed surface through deriveQuestionStatus", () => {
    const callPattern = /\bderiveQuestionStatus\s*\(/;
    for (const relativePath of STATUS_SURFACES) {
      const source = readFileSync(
        path.join(PACKAGE_ROOT, relativePath),
        "utf8",
      );
      expect(callPattern.test(source)).toBe(true);
    }
  });

  it("does not let feature screens map job.state to status keys inline", () => {
    const forbidden = [
      /\bcase\s+["']succeeded["']\s*:/,
      /\bjob\.state\s*===\s*["']succeeded["']/,
      /\bjob\.state\s*===\s*["']failed["']/,
      /\bjob\.state\s*===\s*["']running["']/,
    ];
    for (const relativePath of STATUS_SURFACES.filter((p) =>
      p.includes("/features/"),
    )) {
      const source = readFileSync(
        path.join(PACKAGE_ROOT, relativePath),
        "utf8",
      );
      const offenders = forbidden.filter((pattern) => pattern.test(source));
      expect(offenders).toEqual([]);
    }
  });
});
